In [1]:
import json

with open("clickup_space_flattened_schema.json") as f:
    flattened_schema = json.load(f)
    
flattened_schema

[{'description': 'View the Spaces available in a Workspace.',
  'name': 'get_spaces',
  'parameters': {'properties': {'archived': {'description': 'A flag to decide whether to include archived spaces or not__',
     'type': 'boolean'},
    'team_id': {'description': 'The ID of the team__', 'type': 'string'}},
   'required': ['team_id', 'archived'],
   'type': 'object'}},
 {'description': 'Add a new Space to a Workspace.',
  'name': 'create_space',
  'parameters': {'properties': {'features__due_dates__enabled': {'description': 'enabled__Due dates feature settings__Enabled features within the space__',
     'type': 'boolean'},
    'features__due_dates__remap_closed_due_date': {'description': 'remap_closed_due_date__Due dates feature settings__Enabled features within the space__',
     'type': 'boolean'},
    'features__due_dates__remap_due_dates': {'description': 'remap_due_dates__Due dates feature settings__Enabled features within the space__',
     'type': 'boolean'},
    'features__due

In [2]:
from pprint import pprint

schema_funcs_param = {}

for sc in flattened_schema:
    print(sc["name"], "-", sc["description"])
    single_func_param = {}
    for sck, scv in sc["parameters"]["properties"].items():
        single_func_param[sck] = scv['description']
    schema_funcs_param[sc["name"]] = single_func_param

pprint(schema_funcs_param)  # print(f"\t{sck} - {scv['description']}")

get_spaces - View the Spaces available in a Workspace.
create_space - Add a new Space to a Workspace.
get_space - View the details of a specific Space in a Workspace.
update_space - Rename, set the Space color, and enable ClickApps for a Space.
delete_space - Delete a Space from your Workspace.
get_space_tags - View the task Tags available in a Space.
create_space_tag - Add a new task Tag to a Space.
delete_space_tag - Delete a task Tag from a Space.
{'create_space': {'features__due_dates__enabled': 'enabled__Due dates feature '
                                                  'settings__Enabled features '
                                                  'within the space__',
                  'features__due_dates__remap_closed_due_date': 'remap_closed_due_date__Due '
                                                                'dates feature '
                                                                'settings__Enabled '
                                                     

In [3]:
schema_func_decription_dict = {
    "get_spaces": "Retrives information of all the spaces available in user's Clickup Workspace.",
    "create_space": "Creates a new ClickUp space",
    "get_space": "Retrives information of a specific Clickup space",
    "update_space": "Modifies name, settings the Space color, and assignee management Space.",
    "delete_space": "Delete an existing space from user's ClickUp Workspace",
    "get_space_tags": "Retrives all the Tags assigned on all the tasks in a Space.",
    "create_space_tag": "Assigns a customized Tag in a ClickUp Space.",
    "delete_space_tag": "Deletes a specific tag previously assigned in a space.",
}

schema_func_params_dict = {
    'create_space': {
        'features__due_dates__enabled': 'If due date feature is enabled within the space. \nExample: True, False \nDefault: True',
        'features__due_dates__remap_closed_due_date': 'If remapping closed date feature in due dates is available within the space. \nExample: True, False \nDefault: False',
        'features__due_dates__remap_due_dates': 'If remapping due date feature in due dates is available within the space. \nExample: True, False \nDefault: False',
        'features__due_dates__start_date': 'If start date feature in due dates is available within the space. \nExample: True, False \nDefault: False',
        'features__time_tracking__enabled': 'If time tracking feature is available within the space. \nExample: True, False \nDefault: True',
        'multiple_assignees': 'Enable or disable multiple assignees for tasks within the space \nExample: True, False. Default: True',
        'name': 'The name of the new space to create \nExample:  \'NewWorkspace\', \'TempWorkspace\'',
        'team_id': 'The ID of the team \nExample:  \'abc123\', \'def456\' '
        },
    'create_space_tag': {
        'space_id': 'The ID of the space \nExample:  \'abc123\', \'def456\'',
        'tag__name': 'The name of the tag to assign \nExample:  \'NewTag\', \'TempTag\'',
        'tag__tag_bg': 'The background color of the tag to assign \nExample:  \'#FF0000\', \'#00FF00\'',
        'tag__tag_fg': 'The foreground(text) color of the tag to assign \nExample:  \'#FF0000\', \'#00FF00\''
        },
    'delete_space': {
        'space_id': 'The ID of the space to delete \nExample:  \'abc123\', \'def456\''
        },
    'delete_space_tag': {
        'space_id': 'The ID of the space to delete \nExample:  \'abc123\', \'def456\'',
        'tag__name': 'The name of the tag to delete \nExample:  \'NewTag\', \'TempTag\'',
        'tag__tag_bg': 'The background color of the tag to delete \nExample:  \'#FF0000\', \'#00FF00\', \'#0000FF\'',
        'tag__tag_fg': 'The foreground color of the tag to delete \nExample:  \'#FF0000\', \'#00FF00\', \'#0000FF\'',
        'tag_name': 'The name of the tag to delete \nExample:  \'NewTag\', \'TempTag\''
        },
    'get_space': {
        'space_id': 'The ID of the space to retrieve details \nExample:  \'abc123\', \'def456\''
        },
    'get_space_tags': {
        'space_id': 'The ID of the space to retrieve all the tags from \nExample:  \'abc123\', \'def456\''
        },
    'get_spaces': {
        'archived': 'A flag to decide whether to include archived spaces or not \nExample:  True, False. Default: True',
        'team_id': 'The ID of the team \nExample:  \'abc123\', \'def456\''
        },
    'update_space': {
        'admin_can_manage': 'A flag to determine if the administrator can manage the space or not \nExample:  True, False. Default: True',
        'color': 'The color used for the space \nExample:  \'#FF0000\', \'#00FF00\'',
        'features__due_dates__enabled': 'If due date feature is enabled within the space. \nExample:  True, False \nDefault: True',
        'features__due_dates__remap_closed_due_date': 'If remapping closed date feature in due dates is available within the space. Default: False',
        'features__due_dates__remap_due_dates': 'If remapping due date feature in due dates is available within the space. Default: False',
        'features__due_dates__start_date': 'If start date feature in due dates is available within the space. Default: False',
        'features__time_tracking__enabled': 'If time tracking feature is available within the space. \nExample:  True, False \nDefault: True',
        'multiple_assignees': 'Enable or disable multiple assignees for tasks within the space \nExample:  True, False. Default: True',
        'name': 'The new name of the space \nExample:  \'NewWorkspace\', \'TempWorkspace\'',
        'private': 'A flag to determine if the space is private or not \nExample:  True, False. Default: False',
        'space_id': 'The ID of the space to update \nExample:  \'abc123\', \'def456\''
        }
        }
    
optimized_schema = []
for sc in flattened_schema:
    temp_dict = sc.copy()
    temp_dict["description"] = schema_func_decription_dict[temp_dict["name"]]
    for func_param_name, func_param_description in schema_func_params_dict[temp_dict["name"]].items():
        sc["parameters"]["properties"][func_param_name]["description"] = func_param_description
    optimized_schema.append(temp_dict)

with open('clickup_space_flattened_optimized3_schema.json', 'w') as f:
    json.dump(optimized_schema, f, indent=4, sort_keys=True)
    
optimized_schema

[{'description': "Retrives information of all the spaces available in user's Clickup Workspace.",
  'name': 'get_spaces',
  'parameters': {'properties': {'archived': {'description': 'A flag to decide whether to include archived spaces or not \nExample:  True, False. Default: True',
     'type': 'boolean'},
    'team_id': {'description': "The ID of the team \nExample:  'abc123', 'def456'",
     'type': 'string'}},
   'required': ['team_id', 'archived'],
   'type': 'object'}},
 {'description': 'Creates a new ClickUp space',
  'name': 'create_space',
  'parameters': {'properties': {'features__due_dates__enabled': {'description': 'If due date feature is enabled within the space. \nExample: True, False \nDefault: True',
     'type': 'boolean'},
    'features__due_dates__remap_closed_due_date': {'description': 'If remapping closed date feature in due dates is available within the space. \nExample: True, False \nDefault: False',
     'type': 'boolean'},
    'features__due_dates__remap_due_dat

In [4]:
from pprint import pprint

def flatten_dict(d, parent_key='', sep='__'):
    items = []
    for k, v in d.items():
        new_key = f"{parent_key}{sep}{k}" if parent_key else k
        if isinstance(v, dict):
            items.extend(flatten_dict(v, new_key, sep=sep).items())
        else:
            items.append((new_key, v))
    return dict(items)

# Your original dictionary
original_dict = {
    'features': {
        'due_dates': {
            'enabled': False,
            'remap_closed_due_date': False,
            'remap_due_dates': False,
            'start_date': False
        },
        'time_tracking': {
            'enabled': False
        }
    },
    'multiple_assignees': True,
    'name': 'Innovative Campaigns 2023',
    'team_id': 'cm789'
}

# Flatten the dictionary
flattened_dict = flatten_dict(original_dict)

pprint(flattened_dict)

{'features__due_dates__enabled': False,
 'features__due_dates__remap_closed_due_date': False,
 'features__due_dates__remap_due_dates': False,
 'features__due_dates__start_date': False,
 'features__time_tracking__enabled': False,
 'multiple_assignees': True,
 'name': 'Innovative Campaigns 2023',
 'team_id': 'cm789'}


In [5]:
import json

with open("clickup_space_benchmark.json") as f:
    bench_data = json.load(f)
    
bench_data[1]

{'prompt': 'Ella, the project coordinator, is setting up a new project space in ClickUp for the "Creative Minds" team with team ID "cm789". This space, named "Innovative Campaigns 2023", should allow multiple assignees for tasks, but keep due dates and time tracking disabled, as the initial planning phase doesn\'t require strict deadlines or time monitoring.',
 'solution': 'create_space(team_id="cm789", name="Innovative Campaigns 2023", multiple_assignees=True, features=(due_dates=(enabled=False, start_date=False, remap_due_dates=False, remap_closed_due_date=False), time_tracking=(enabled=False)))'}

In [6]:
import os
import json
import dotenv
from textwrap import dedent
from datetime import datetime

dotenv.load_dotenv()

tools = [{"name":sc["name"],  "description":sc["description"], "input_schema": sc["parameters"]} for sc in optimized_schema]
tools

[{'name': 'get_spaces',
  'description': "Retrives information of all the spaces available in user's Clickup Workspace.",
  'input_schema': {'properties': {'archived': {'description': 'A flag to decide whether to include archived spaces or not \nExample:  True, False. Default: True',
     'type': 'boolean'},
    'team_id': {'description': "The ID of the team \nExample:  'abc123', 'def456'",
     'type': 'string'}},
   'required': ['team_id', 'archived'],
   'type': 'object'}},
 {'name': 'create_space',
  'description': 'Creates a new ClickUp space',
  'input_schema': {'properties': {'features__due_dates__enabled': {'description': 'If due date feature is enabled within the space. \nExample: True, False \nDefault: True',
     'type': 'boolean'},
    'features__due_dates__remap_closed_due_date': {'description': 'If remapping closed date feature in due dates is available within the space. \nExample: True, False \nDefault: False',
     'type': 'boolean'},
    'features__due_dates__remap_due

In [7]:
import anthropic
client = anthropic.Anthropic()


tool_trailer = "\n".join([f'{sc["name"]} - {sc["description"]}' for sc in flattened_schema])

fcalling_llm = lambda fprompt : client.beta.tools.messages.create(
  model="claude-3-sonnet-20240229",
  system =  f"""
You are an agent who is responsible for managing various employee management platform, 
one of which is CliuckUp. 

You are given a number of tools as functions, you must use one of those tools and fillup 
all the parameters of those tools ,whose answers you will get from the given situation.

When you are presented with a technical situation, that a person of a team is facing, 
you must give the soulution utilizing your functionalities. 

The Functions provided to you are:
{tool_trailer}

First analyze the given situation to fully anderstand what is the intention of the user,
what they need and exactly which tool will fill up that necessity.

Then look into the parameters and extract all the relevant informations to fillup the 
parameter with right values.
""",
  messages=[
    {
      "role": "user",
      "content": fprompt
    },
  ],
  temperature=0,
  max_tokens=4096,
  top_p=1,
  tools=tools,
)

response = fcalling_llm(bench_data[1]["prompt"])
# response
response

ToolsBetaMessage(id='msg_013n55P91m3aQxjaemGEEXfZ', content=[TextBlock(text="Okay, let's use the ", type='text'), ToolUseBlock(id='toolu_01FwR1HHRZyrkmCFwCTx99hi', input={'team_id': 'cm789', 'name': 'Innovative Campaigns 2023', 'multiple_assignees': True, 'features__due_dates__enabled': False, 'features__due_dates__start_date': False, 'features__due_dates__remap_due_dates': False, 'features__due_dates__remap_closed_due_date': False, 'features__time_tracking__enabled': False}, name='create_space', type='tool_use')], model='claude-3-haiku-20240307', role='assistant', stop_reason='tool_use', stop_sequence=None, type='message', usage=Usage(input_tokens=2640, output_tokens=235))

In [8]:
import re 
from pprint import pprint



class FuncStringParser():
    def __init__(self):
        self.record_keeper = {}
        
        
    def replace_with_dict(self, match):
        replacement = f"dict{len(self.record_keeper)}"
        self.record_keeper[replacement] = match.group()
        # dict_counter += 1
        return replacement

    def arg_tuple_str_to_args_dict(self, func_args):
        # print(func_args)
        args_dict = {}
        func_args = func_args[1:-1].split(",")
        for arg in func_args:
            k, v = arg[:arg.find("=")], arg[arg.find("=")+1:]
            k, v = k.strip(), v.strip()
            
            if v.startswith("(") and v.endswith(")"):
                v = self.func_to_args_dict(v)
                # print(v)
            
            args_dict[k] = v

        return args_dict

    def replace_true_false_dfs(self, nested_dict):
        for key, value in nested_dict.items():
            # If the value is a nested dictionary, recursively process it
            if isinstance(value, dict):
                self.replace_true_false_dfs(value)
            else:
                nested_dict[key] = eval(value)
                
                    
    def func_to_args_dict(self, func_args):
        pattern_uncompiled = r'\(([^()]+)\)'
        pattern = re.compile(pattern_uncompiled)
        # print(re.findall(pattern_uncompiled, func_args))
        while len(re.findall(pattern_uncompiled, func_args)) > 1:
            func_args = re.sub(pattern, self.replace_with_dict, func_args)
            
        # print(re.findall(pattern_uncompiled, func_args))
        if sum([m.start(0) > 2 for m in re.finditer(r'\(([^()]+)\)', func_args)]):
            func_args = re.sub(pattern, self.replace_with_dict, func_args)
            # print(func_args)
        func_args = re.sub(pattern, self.replace_with_dict, func_args)
        # pprint(func_args)
            
        dict_record_keeper = {k:self.arg_tuple_str_to_args_dict(v) for k, v in self.record_keeper.items()}
        # print(dict_record_keeper)
        
        # inv_dict_record_keeper = {v:k for k,v in dict_record_keeper.items()}
        for i1, (k1, v1) in enumerate(dict_record_keeper.items()):
            for i2, (k2, v2) in enumerate(dict_record_keeper.items()):
                if i2 == i1:
                    break
                if k2 in v1.values():
                    # print(k2, v1)
                    dict_record_keeper[k1][list(v1.keys())[list(v1.values()).index(k2)]] = v2
        args_dict = list(dict_record_keeper.values())[-1]
        self.replace_true_false_dfs(args_dict)
        self.record_keeper = {}
        return args_dict
    
    def func_string_to_func_dict(self, func_string):
        func_name = func_string[:func_string.find("(")]
        func_args = func_string[func_string.find("("):]
        func_args_dict = self.func_to_args_dict(func_args)
        return {"name": func_name, "arguments": func_args_dict}
    
    def llm_response_to_func_dict(self, llm_response):
        for message in llm_response.content:
            if message.type == "tool_use":
                return {
                    "name": message.name,
                    "arguments": message.input,
                }
        return {"name": None, "arguments": None}
        

                    
        
    
# Sample string
func = 'create_space(team_id="cm789", name="Innovative Campaigns 2023", multiple_assignees=True, features=(due_dates=(enabled=False, start_date=False, remap_due_dates=False, remap_closed_due_date=False), time_tracking=(enabled=False)))'

func_name = func[:func.find("(")]
func_args = func[func.find("("):]

fsp = FuncStringParser()
# fsp.func_to_args_dict(func_args)

fsp.func_string_to_func_dict(func)


{'name': 'create_space',
 'arguments': {'team_id': 'cm789',
  'name': 'Innovative Campaigns 2023',
  'multiple_assignees': True,
  'features': {'due_dates': {'enabled': False,
    'start_date': False,
    'remap_due_dates': False,
    'remap_closed_due_date': False},
   'time_tracking': {'enabled': False}}}}

In [9]:
from tqdm import tqdm
# from tqdm.auto import tqdm


llm_output = []

for bdata in tqdm(bench_data):
    response = fcalling_llm(bdata["prompt"])
    try:
        llm_output.append(response)
    except TypeError:
        print(bdata["prompt"])
        print(response)
        llm_output.append(None)
        print("------")
llm_output

  0%|          | 0/50 [00:00<?, ?it/s]

100%|██████████| 50/50 [01:48<00:00,  2.16s/it]


[ToolsBetaMessage(id='msg_01TkXRaNMKKLBQaRDd1yP42D', content=[TextBlock(text="Okay, let's use the ", type='text'), ToolUseBlock(id='toolu_01RjhU3cEXqVAzvaSBE9vwCM', input={'team_id': 'team123', 'archived': True}, name='get_spaces', type='tool_use')], model='claude-3-haiku-20240307', role='assistant', stop_reason='tool_use', stop_sequence=None, type='message', usage=Usage(input_tokens=2634, output_tokens=80)),
 ToolsBetaMessage(id='msg_0172PgRuZ2SP5GxUQQkXAKVN', content=[TextBlock(text="Okay, let's use the ", type='text'), ToolUseBlock(id='toolu_01NZXHdvqwYV9Q1VYPirCbPc', input={'team_id': 'cm789', 'name': 'Innovative Campaigns 2023', 'multiple_assignees': True, 'features__due_dates__enabled': False, 'features__due_dates__start_date': False, 'features__due_dates__remap_due_dates': False, 'features__due_dates__remap_closed_due_date': False, 'features__time_tracking__enabled': False}, name='create_space', type='tool_use')], model='claude-3-haiku-20240307', role='assistant', stop_reason='t

In [10]:
from pprint import pprint

fsp = FuncStringParser()
result_list = []

# fsp.func_to_args_dict(func_args)
for bdata, llm_response in zip(bench_data, llm_output):
    bdata_solution_dict = fsp.func_string_to_func_dict(bdata["solution"])
    # bdata_solution_dict["name"] = schema_func_name_dict[bdata_solution_dict["name"]]
    bdata_solution_dict["arguments"] = flatten_dict(bdata_solution_dict["arguments"])
    
    llm_response_dict = fsp.llm_response_to_func_dict(llm_response)
    
    pprint(bdata_solution_dict)
    pprint(llm_response_dict)
    print(llm_response_dict == bdata_solution_dict)
    result_list.append(llm_response_dict == bdata_solution_dict)
    print("-----")
    
run_acc_1 = sum(result_list)/len(result_list)
run_acc_1

{'arguments': {'archived': False, 'team_id': 'team123'}, 'name': 'get_spaces'}
{'arguments': {'archived': True, 'team_id': 'team123'}, 'name': 'get_spaces'}
False
-----
{'arguments': {'features__due_dates__enabled': False,
               'features__due_dates__remap_closed_due_date': False,
               'features__due_dates__remap_due_dates': False,
               'features__due_dates__start_date': False,
               'features__time_tracking__enabled': False,
               'multiple_assignees': True,
               'name': 'Innovative Campaigns 2023',
               'team_id': 'cm789'},
 'name': 'create_space'}
{'arguments': {'features__due_dates__enabled': False,
               'features__due_dates__remap_closed_due_date': False,
               'features__due_dates__remap_due_dates': False,
               'features__due_dates__start_date': False,
               'features__time_tracking__enabled': False,
               'multiple_assignees': True,
               'name': 'Innovative

0.68

In [11]:
from tqdm import tqdm
# from tqdm.auto import tqdm


llm_output = []

for bdata in tqdm(bench_data):
    response = fcalling_llm(bdata["prompt"])
    try:
        llm_output.append(response)
    except TypeError:
        print(bdata["prompt"])
        print(response)
        llm_output.append(None)
        print("------")
llm_output

100%|██████████| 50/50 [01:58<00:00,  2.37s/it]


[ToolsBetaMessage(id='msg_016UGSCu7wb8mR1PQaTJPVRE', content=[TextBlock(text="Okay, let's use the ", type='text'), ToolUseBlock(id='toolu_01JGkTQVguCTag9eQuAF9eof', input={'team_id': 'team123', 'archived': True}, name='get_spaces', type='tool_use')], model='claude-3-haiku-20240307', role='assistant', stop_reason='tool_use', stop_sequence=None, type='message', usage=Usage(input_tokens=2634, output_tokens=80)),
 ToolsBetaMessage(id='msg_01Kgyb3XHhUYTZxvtBz1wDnk', content=[TextBlock(text="Okay, let's use the ", type='text'), ToolUseBlock(id='toolu_011hYTL2pMkGEnbLLdR7cMAD', input={'team_id': 'cm789', 'name': 'Innovative Campaigns 2023', 'multiple_assignees': True, 'features__due_dates__enabled': False, 'features__due_dates__start_date': False, 'features__due_dates__remap_due_dates': False, 'features__due_dates__remap_closed_due_date': False, 'features__time_tracking__enabled': False}, name='create_space', type='tool_use')], model='claude-3-haiku-20240307', role='assistant', stop_reason='t

In [12]:
from pprint import pprint

fsp = FuncStringParser()
result_list = []

# fsp.func_to_args_dict(func_args)
for bdata, llm_response in zip(bench_data, llm_output):
    bdata_solution_dict = fsp.func_string_to_func_dict(bdata["solution"])
    # bdata_solution_dict["name"] = schema_func_name_dict[bdata_solution_dict["name"]]
    bdata_solution_dict["arguments"] = flatten_dict(bdata_solution_dict["arguments"])
    
    llm_response_dict = fsp.llm_response_to_func_dict(llm_response)
    
    pprint(bdata_solution_dict)
    pprint(llm_response_dict)
    print(llm_response_dict == bdata_solution_dict)
    result_list.append(llm_response_dict == bdata_solution_dict)
    print("-----")
    
run_acc_2 = sum(result_list)/len(result_list)
run_acc_2

{'arguments': {'archived': False, 'team_id': 'team123'}, 'name': 'get_spaces'}
{'arguments': {'archived': True, 'team_id': 'team123'}, 'name': 'get_spaces'}
False
-----
{'arguments': {'features__due_dates__enabled': False,
               'features__due_dates__remap_closed_due_date': False,
               'features__due_dates__remap_due_dates': False,
               'features__due_dates__start_date': False,
               'features__time_tracking__enabled': False,
               'multiple_assignees': True,
               'name': 'Innovative Campaigns 2023',
               'team_id': 'cm789'},
 'name': 'create_space'}
{'arguments': {'features__due_dates__enabled': False,
               'features__due_dates__remap_closed_due_date': False,
               'features__due_dates__remap_due_dates': False,
               'features__due_dates__start_date': False,
               'features__time_tracking__enabled': False,
               'multiple_assignees': True,
               'name': 'Innovative

0.68

In [13]:
from tqdm import tqdm
# from tqdm.auto import tqdm


llm_output = []

for bdata in tqdm(bench_data):
    response = fcalling_llm(bdata["prompt"])
    try:
        llm_output.append(response)
    except TypeError:
        print(bdata["prompt"])
        print(response)
        llm_output.append(None)
        print("------")
llm_output

100%|██████████| 50/50 [01:52<00:00,  2.25s/it]


[ToolsBetaMessage(id='msg_01U87sZe6e2cfxoZvvZTHYxP', content=[TextBlock(text="Okay, let's use the ", type='text'), ToolUseBlock(id='toolu_01UBg4D2XkWXGkPBMs9GaK9y', input={'team_id': 'team123', 'archived': True}, name='get_spaces', type='tool_use')], model='claude-3-haiku-20240307', role='assistant', stop_reason='tool_use', stop_sequence=None, type='message', usage=Usage(input_tokens=2634, output_tokens=80)),
 ToolsBetaMessage(id='msg_0182MarMUye4sHJoCXWhnn6R', content=[TextBlock(text="Okay, let's use the ", type='text'), ToolUseBlock(id='toolu_01Ce8y1YNXia9JZ7rtWhGnG2', input={'team_id': 'cm789', 'name': 'Innovative Campaigns 2023', 'multiple_assignees': True, 'features__due_dates__enabled': False, 'features__due_dates__start_date': False, 'features__due_dates__remap_due_dates': False, 'features__due_dates__remap_closed_due_date': False, 'features__time_tracking__enabled': False}, name='create_space', type='tool_use')], model='claude-3-haiku-20240307', role='assistant', stop_reason='t

In [14]:
from pprint import pprint

fsp = FuncStringParser()
result_list = []

# fsp.func_to_args_dict(func_args)
for bdata, llm_response in zip(bench_data, llm_output):
    bdata_solution_dict = fsp.func_string_to_func_dict(bdata["solution"])
    bdata_solution_dict["arguments"] = flatten_dict(bdata_solution_dict["arguments"])
    
    llm_response_dict = fsp.llm_response_to_func_dict(llm_response)
    
    pprint(bdata_solution_dict)
    pprint(llm_response_dict)
    print(llm_response_dict == bdata_solution_dict)
    result_list.append(llm_response_dict == bdata_solution_dict)
    print("-----")
    
run_acc_3 = sum(result_list)/len(result_list)
run_acc_3

{'arguments': {'archived': False, 'team_id': 'team123'}, 'name': 'get_spaces'}
{'arguments': {'archived': True, 'team_id': 'team123'}, 'name': 'get_spaces'}
False
-----
{'arguments': {'features__due_dates__enabled': False,
               'features__due_dates__remap_closed_due_date': False,
               'features__due_dates__remap_due_dates': False,
               'features__due_dates__start_date': False,
               'features__time_tracking__enabled': False,
               'multiple_assignees': True,
               'name': 'Innovative Campaigns 2023',
               'team_id': 'cm789'},
 'name': 'create_space'}
{'arguments': {'features__due_dates__enabled': False,
               'features__due_dates__remap_closed_due_date': False,
               'features__due_dates__remap_due_dates': False,
               'features__due_dates__start_date': False,
               'features__time_tracking__enabled': False,
               'multiple_assignees': True,
               'name': 'Innovative

0.68

In [15]:
print("ClickUp Benchmark - Flattened Schema + Focused System Prompt containing Schema summary + \nFunction and Parameter Descriptions Optimized + Function Parameter examples added : Accuracy")
print("=========================================")
print(f"Run 1: {run_acc_1}")
print(f"Run 2: {run_acc_2}")
print(f"Run 3: {run_acc_3}")

ClickUp Benchmark - Flattened Schema + Focused System Prompt containing Schema summary + 
Function and Parameter Descriptions Optimized + Function Parameter examples added : Accuracy
Run 1: 0.68
Run 2: 0.68
Run 3: 0.68


| Approach | Run 1 | Run 2  | Run 3  | Average Accuracy  |
|---|---|---|---|---|
| No System Prompt  | 0.36  | 0.36  | 0.36  | 0.36  |
| Flattening Schema  | 0.54  | 0.5  | 0.54  | 0.527  |
| Flattened Schema + <br>Simple System Prompt  | 0.56  | 0.56  | 0.54  | 0.553  |
| Flattened Schema + <br>Focused System Prompt  | 0.64  | 0.64  | 0.62  | 0.633  |
| Flattened Schema + <br>Focused System Prompt containing Schema summary  | 0.64  | 0.62  | 0.66  | 0.64  |
| Flattened Schema + <br>Focused System Prompt containing Schema summary + <br>Function Name Optimized  | 0.72  | 0.72  | 0.66  | 0.70  |
| Flattened Schema + <br>Focused System Prompt containing Schema summary + <br>Function Description Optimized  |  0.68  | 0.7  | 0.68  | 0.687  |
| Flattened Schema + <br>Focused System Prompt containing Schema summary + <br>Function and Parameter Descriptions Optimized  | 0.78  | 0.76  | 0.76  | 0.767  |
| Flattened Schema + <br>Focused System Prompt containing Schema summary + <br>Function and Parameter Descriptions Optimized + <br>Function Call examples added  | 0.72  | 0.7  | 0.66  | 0.693  |
| Flattened Schema + <br>Focused System Prompt containing Schema summary + <br>Function and Parameter Descriptions Optimized + <br>Function Parameter examples added  | 0.82  | 0.76  | 0.78  | 0.787  |
